### PR TITLE
ELEVATOR MUSIC: True Elevator Experience

### PR DESCRIPTION
**Pull request by DrDiasyl**
## About The Pull Request
Adds elevator music into the game that is played by an elevator panel.

https://github.com/tgstation/tgstation/assets/42353186/1a801604-3990-46ae-a96a-b3766b102d62

It's done by using loop sound, with a Kevin MacLeod "Local Forecast - Elevator" (UNDER CC ATTRIBUTIONS 4.0, and we anyway used some other Kevin MacLeod music) chopped into 8 small pieces.
The elevator panel has a variable which allows playing music but can be changed in the map editor if you don't want it to play at certain places.

(It also doesn't ignore walls, this means you can't hear the music through wall or when elevator is closed)
## Why It's Good For The Game
Gives elevators more flavour and love, especially when people mostly prefer stairs to those "laggy crushing machines."
Because of this people might instead hop into an elevator just to hear meme elevator music, which is relaxing and might create comedic situations (although elevators don't move that fast)
## Changelog
:cl: DrDiasyl aka DrTuxedo
sound: Nanotrasen have started installing music players in the elevators to boost workers' morale.
/:cl:

### CODE REVIEW

File: code/datums/looping_sounds/music.dm [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+	 'sound/ambience/music/local_forecast/local_forecast5.ogg' = 3,
+	 'sound/ambience/music/local_forecast/local_forecast6.ogg' = 2,
+	 'sound/ambience/music/local_forecast/local_forecast7.ogg' = 1,
+	)
+	mid_length = 5.83 SECONDS
+
+	end_sound = 'sound/ambience/music/local_forecast/local_forecast8.ogg'
+
+	volume = 20
+	falloff_exponent = 10
```
[the remark this answers is not in the document]
Comment (DrDiasyl) [reply code/datums/looping_sounds/music.dm:18]: > why is this so STRONG

Oh wait, are you about falloff? I just don't want music to be loud sounding in hallways, that's it

File: code/modules/industrial_lift/elevator/elevator_panel.dm [hunk 2]
```diff
@@ -64,6 +69,9 @@
 	// this is just here for redundancy's sake.
 	. = INITIALIZE_HINT_LATELOAD
 
+	if(elevator_music_toggle)
+		elevator_music = new(src, TRUE)
```
Comment (Mothblocks) [inline code/modules/industrial_lift/elevator/elevator_panel.dm:73]:
```suggestion
		elevator_music = new(src, /* name_of_paremeter_here = */ TRUE)
```

**Comment by TheSmallBlue** [comment]
Hearing that specific Kevin McLeod song brings me back to watching IHE in like 2016

File: code/datums/looping_sounds/music.dm (hunk 1 above)
[the remark this answers is not in the document]
Comment (DrDiasyl) [reply code/datums/looping_sounds/music.dm:18]: Okay I lowered it to 5

File: code/modules/industrial_lift/elevator/elevator_panel.dm (hunk 2 above)
Comment (DrDiasyl) [reply code/modules/industrial_lift/elevator/elevator_panel.dm:73]: @Mothblocks what to do if I have no idea? Just from the SM code I got that it must be that way, for some reason

Comment (DrDiasyl) [reply code/modules/industrial_lift/elevator/elevator_panel.dm:73]: I think it's like "Is it turned on or off?" as initialization of shower has it FALSE

Comment (DrDiasyl) [reply code/modules/industrial_lift/elevator/elevator_panel.dm:73]: From what I see it doesn't really have a name so yeah

Comment (Mothblocks) [reply code/modules/industrial_lift/elevator/elevator_panel.dm:73]: Uh, it definitely has a name.

Look in `/datum/sound_loop/New` and see what the parameter is called, then copy that name.

Comment (DrDiasyl) [reply code/modules/industrial_lift/elevator/elevator_panel.dm:73]: > Uh, it definitely has a name.
> 
> Look in `/datum/sound_loop/New` and see what the parameter is called, then copy that name.

Why every other can just have "TRUE" and "FALSE"? Whatever.
I will find it, if I won't then bleep

**Comment by iwishforducks** [comment]
the elevator roblox real

**Comment by MrMelbert** [comment]
Does the music turn off if the panel loses power? 

~~Can the panel even lose power I forgot if I coded that~~

File: code/modules/industrial_lift/elevator/elevator_panel.dm (hunk 2 above)
Comment (Mothblocks) [reply code/modules/industrial_lift/elevator/elevator_panel.dm:73]: They're not supposed to. It's hard to globally enforce code standards. We're working on it.

All you have to do is search `/datum/sound_loop/New` and figure out what the first parameter is.

**Comment by Farquaar** [comment]
I would love to hear Summertime Samba in the elevators if we can find a CC performance of it.

**Comment by DrDiasyl** [comment]
> Does the music turn off if the panel loses power?
> 
> ~Can the panel even lose power I forgot if I coded that~

I should probably make it turn off when the elevator loses power.

File: code/modules/industrial_lift/elevator/elevator_panel.dm (hunk 2 above)
Comment (DrDiasyl) [reply code/modules/industrial_lift/elevator/elevator_panel.dm]: D o n e

File: code/modules/industrial_lift/elevator/elevator_panel.dm [hunk 3]
```diff
@@ -64,6 +69,9 @@
 	// this is just here for redundancy's sake.
 	. = INITIALIZE_HINT_LATELOAD
 
+	if(elevator_music_toggle)
+		elevator_music = new(src, start_immediately = TRUE)
```
Comment (Mothblocks) [inline code/modules/industrial_lift/elevator/elevator_panel.dm:73]: I'm surprised this works without that comment trick, usually this completely borks with `new`.

File: code/modules/industrial_lift/elevator/elevator_panel.dm [hunk 4]
```diff
@@ -112,14 +120,22 @@
...
 
 	playsound(src, SFX_SPARKS, 100, TRUE, SHORT_RANGE_SOUND_EXTRARANGE)
 	balloon_alert(user, "safeties overridden")
 
+/obj/machinery/elevator_control_panel/power_change()
+	if(machine_stat & (BROKEN|NOPOWER|EMPED))
+		elevator_music.stop()
+	else
+		elevator_music.start()
+	. = ..()
```
Comment (Mothblocks) [inline code/modules/industrial_lift/elevator/elevator_panel.dm:136]:
```suggestion
	return ..()
```

**Review by Mothblocks — APPROVED** [review]

**Comment by BlueMemesauce** [comment]
Shouldn't it only play when the elevator is moving?

**Comment by DrDiasyl** [comment]
> Shouldn't it only play when the elevator is moving? Elevators usually don't have the doors stay open with music constantly playing.  

With how fast elevators move, we wouldn't hear that much of music

File: code/modules/industrial_lift/elevator/elevator_panel.dm (hunk 3 above)
Comment (DrDiasyl) [reply code/modules/industrial_lift/elevator/elevator_panel.dm:73]: It just does 🤷

**Pull request merged**